walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,186 @@
 public final class FindMeetingQuery {
+
+  /**
+   * Finds all time intervals, if they exist, where all a meeting request's mandatory attendees are
+   * not scheduled to attend any other meeting. The set of valid times is returned as an ordered
+   * collection of TimeRanges.
+   *
+   * @param events the collection of all scheduled events
+   * @param request the meeting request
+   * @return an ordered collection of TimeRanges where all attendees can attend a meeting, or an
+   *     empty collection if no valid times exist.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Expected O(M) time
+    Set<String> requestedAttendees = new HashSet<String>(request.getAttendees());
+
+    // Expected O(E log E) time
+    SortedMap<Integer, TimeRange> rawConflictSet =
+        generateRawConflictSet(events, requestedAttendees);
+    Collection<TimeRange> conflictSet = cleanConflictSet(rawConflictSet.values());
+    return getValidTimes(conflictSet, request.getDuration());
+  }
+
+  /**
+   * Algorithm design approach:
+   *
+   * <p>- To build the set of valid meeting times, I find the set of invalid meeting times and take
+   * the complement of that set.
+   *
+   * <p>- The following is a sketch of how I build the set of invalid meeting times:
+   *
+   * <pre>
+   * S = the empty set
+   * for each m in the set of all meetings,
+   *   if INTERSECT(request's attendees, m's attendees) is non-empty, (1)
+   *     S := UNION(S, m's time range) (2)
+   * </pre>
+   *
+   * <p>- I choose an appropriate implementation for each of the abstract operations above. (1) is
+   * implemented as a hashtable lookup, where each of m's attendees is looked up in the prepared
+   * hashtable of request's attendees, in O(m's attendees) time. (2) is implemented in two phases,
+   * the first being the insertion of all conflicting time ranges into a set ordered by start time,
+   * with O(log n) insertion time. The first phase returns a set which may have several overlapping
+   * entries. Thus, the second phase, done outside of the loop, converts the raw set in O(n) time
+   * into a set without overlapping entries.
+   */
+
+  /**
+   * Generates a set of time ranges that conflict with any of the requested attendees' schedules.
+   * Note that ranges may overlap. Expected runtime is O(E log E).
+   *
+   * @param events the collection of all events
+   * @param requestedAttendees the set of all requested attendees.
+   * @return a map of potentially overlapping time ranges, with the start time as the key
+   */
+  private SortedMap<Integer, TimeRange> generateRawConflictSet(
+      Collection<Event> events, Set<String> requestedAttendees) {
+    SortedMap<Integer, TimeRange> rawConflictSet = new TreeMap<Integer, TimeRange>();
+    for (Event event : events) {
+      if (attendeesIntersect(event.getAttendees(), requestedAttendees)) {
```

Comment:
You can flip this to save an indentation:
```java
if (!attendeesIntersect(...)) {
  continue;
}
/** other logic **/
```

Reply:
I'd prefer avoiding the `continue` statement. Though I could see making this change if the `if` statement body were large.

Looking at this snippet again, I have a small refactor to make the body slightly smaller.